jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -306,21 +311,48 @@ private void validateMemScope(ResourceScope memScope) throws IllegalStateExcepti
 	 * of the passed-in pointer argument to long.
 	 */
 	private final long memSegmtOfPtrToLongArg(MemorySegment argValue, LinkerOptions options) throws IllegalStateException {
-		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
-		addMemArgScope(argValue.scope());
+		try {
+			UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
+			addMemArgScope(argValue.scope());
+		} catch (Throwable e) {
+			/*[IF JAVA_SPEC_VERSION >= 22]*/
+			/* Notify to reset the heap argument information for the current thread
+			 * due to the exception triggered here to guarantee the same handler with
+			 * different heap arguments passes through next time.
+			 */
+			HeapArgInfo info = heapArgInfo.get();
```

Comment:
Suggest moving this declaration to the beginning of the method so we only have one copy of the code to fetch it:
```java
		/*[IF JAVA_SPEC_VERSION >= 22]*/
		HeapArgInfo info = heapArgInfo.get();
		/*[ENDIF] JAVA_SPEC_VERSION >= 22 */
```

Reply:
Fixed.